contracts/CollectiveGovernance.sol

```diff
@@ -94,59 +95,66 @@ contract CollectiveGovernance is Governance, VoteStrategy, ERC165 {
     /// @param _governanceStorage The storage contract for this governance
     /// @param _gasUsedRebate The maximum rebate for gas used
     /// @param _baseFeeRebate The maximum base fee rebate
-    ///
     constructor(
         address[] memory _supervisorList,
         VoterClass _class,
         Storage _governanceStorage,
+        MetaStorage _metaStore,
         TimeLocker _timeLocker,
         uint256 _gasUsedRebate,
         uint256 _baseFeeRebate
     ) {
-        require(_supervisorList.length > 0, "Supervisor required");
-        require(_gasUsedRebate >= Constant.MAXIMUM_REBATE_GAS_USED, "Insufficient gas used rebate");
-        require(_baseFeeRebate >= Constant.MAXIMUM_REBATE_BASE_FEE, "Insufficient base fee rebate");
-
+        if (_supervisorList.length == 0) revert SupervisorListEmpty();
+        if (_gasUsedRebate < Constant.MAXIMUM_REBATE_GAS_USED)
+            revert GasUsedRebateMustBeLarger(_gasUsedRebate, Constant.MAXIMUM_REBATE_GAS_USED);
+        if (_baseFeeRebate < Constant.MAXIMUM_REBATE_BASE_FEE)
+            revert BaseFeeRebateMustBeLarger(_baseFeeRebate, Constant.MAXIMUM_REBATE_BASE_FEE);
         _voterClass = _class;
         _storage = _governanceStorage;
+        meta = _metaStore;
         _timeLock = _timeLocker;
-        _communitySupervisorList = _supervisorList;
         _maximumGasUsedRebate = _gasUsedRebate;
         _maximumBaseFeeRebate = _baseFeeRebate;
+        _communitySupervisorList = _supervisorList;
+    }
+
+    modifier requireNotFinal(uint256 _proposalId) {
+        if (_storage.isFinal(_proposalId)) revert VoteFinal(_proposalId);
+        _;
     }
 
-    modifier requireVoteReady(uint256 _proposalId) {
-        require(_storage.isFinal(_proposalId), "Voting is not ready");
+    modifier requireVoteFinal(uint256 _proposalId) {
+        if (!_storage.isFinal(_proposalId)) revert VoteNotFinal(_proposalId);
         _;
     }
 
     modifier requireVoteClosed(uint256 _proposalId) {
-        require(_storage.isFinal(_proposalId) && !isVoteOpenByProposalId[_proposalId], "Vote is not closed");
+        if (isVoteOpenByProposalId[_proposalId]) revert VoteIsOpen(_proposalId);
         _;
     }
 
     modifier requireVoteOpen(uint256 _proposalId) {
-        require(_storage.isFinal(_proposalId) && isVoteOpenByProposalId[_proposalId], "Voting is closed");
+        if (!isVoteOpenByProposalId[_proposalId]) revert VoteIsClosed(_proposalId);
```

Comment:
just curious what the difference is between failing via require and with revert

Reply:
Custom errors were introduced in Solidity 0.8.4, https://blog.soliditylang.org/2021/04/21/custom-errors/.   These have many advantages.  First strings in solidity are problematic because of size limits.  Also from a programming POV each string is pretty meaningless and depends alot on the English language.  This is also bad with the size limits because you get forced to make incredibly short and terse messages that are even more meaningless.   However with custom errors the exact error and its parameters can be expected and tested for (grep for abi.encodeWithSelector in the tests).   This makes it easier to prove that we have 1 test for each branch in the code.   And the meaning of each custom error should be obvious for that branch.  So, TLDR, smaller in size and easier to test.